doc/man/ovdb_init.8
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "OVDB_INIT 8"
.TH OVDB_INIT 8 "2008-04-06" "INN 2.4.6" "InterNetNews Documentation"
.SH "NAME"
ovdb_init \- Prepare ovdb database for use
.SH "SYNOPSYS"
.IX Header "SYNOPSYS"
ovdb_init [\f(CW\*(C`\-u\*(C'\fR|\f(CW\*(C`\-r\*(C'\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This command must be run before any other process can access the
overview database.  It performs the following steps:
.IP "1" 4
.IX Item "1"
Creates the database environment, if necessary
.IP "2" 4
.IX Item "2"
If the database is idle (and if the \f(CW\*(C`\-u\*(C'\fR option is not specified),
it performs a normal recovery.  The recovery will remove stale locks,
recreate the memory pool cache, and repair any damage caused by a system
crash or improper shutdown.
.IP "3" 4
.IX Item "3"
If the \f(CW\*(C`\-u\*(C'\fR option is specified, it performs any necessary upgrades
to the database.  See the \s-1UPGRADING\s0 section below.
.IP "4" 4
.IX Item "4"
Starts the \s-1DB\s0 housekeeping processes (ovdb_monitor) if they're not
already running. (Unless the \f(CW\*(C`\-r\*(C'\fR option is specified).
.IP "5" 4
.IX Item "5"
Starts the ovdb readserver (ovdb_server) processes if \f(CW\*(C`readserver\*(C'\fR
in \fIovdb.conf\fR is \f(CW\*(C`true\*(C'\fR, and if they're not
already running. (Unless the \f(CW\*(C`\-r\*(C'\fR option is specified).
.PP
Returns exit status of 0 if all steps were completed successfully.
In the event of an error, messages are written to syslog and/or stderr.
.PP
If a recovery was attempted but it failed, the database may be
damaged beyond repair, requiring a rebuild with \fImakehistory\fR\|(8).
.PP
This command is normally invoked automatically by \fIrc.news\fR\|(8).
.PP
It is \s-1OK\s0 to run this command multiple times.
.SH "OPTIONS"
.IX Header "OPTIONS"
.ie n .IP """\-r""" 4
.el .IP "\f(CW\-r\fR" 4
.IX Item "-r"
Perform recovery only.  \f(CW\*(C`ovdb_monitor\*(C'\fR is not started.
.ie n .IP """\-u""" 4
.el .IP "\f(CW\-u\fR" 4
.IX Item "-u"
Perform any needed upgrades.  Recovery is not attempted.
\&\f(CW\*(C`ovdb_monitor\*(C'\fR is started if the upgrade succeeded.
.SH "UPGRADING"
.IX Header "UPGRADING"
There are two situations in which the database will need to be
upgraded:
.IP "\(bu" 4
You upgrade the BerkeleyDB library to a newer version, for example
from 2.7.7 to 3.1.17.  In this case, the BerkeleyDB db\->\fIupgrade()\fR
method is used.
.IP "\(bu" 4
You upgrade ovdb to a newer major version; i.e., ovdb\-1.0 to ovdb\-2.0.
.PP
In both of these cases, the database is upgraded in\-place; and the
upgrade can not be undone.  Do not interrupt the upgrade process once
it has started, because there is a risk of irrepairable corruption.
The upgrade may take several minutes to complete.
If an upgrade does get interrupted, try running the upgrade again.
.PP
Here's an example procedure to upgrade a database created with BerkeleyDB
2.7.7 to use BerkeleyDB 3.1.17:
.IP "1" 4
.IX Item "1"
Build and install the BerkeleyDB 3.1.17
.IP "2" 4
.IX Item "2"
Run configure in the \s-1INN\s0 source tree and make sure it picks up the
right BerkeleyDB directory (e.g., /usr/local/BerkeleyDB.3.1)
.IP "3" 4
.IX Item "3"
Do a \f(CW\*(C`make\*(C'\fR
.IP "4" 4
.IX Item "4"
Shut down \s-1INN\s0 (e.g., with \f(CW\*(C`rc.news stop\*(C'\fR).  Be sure to kill all nnrpds as
well.
.IP "5" 4
.IX Item "5"
Do a \f(CW\*(C`make update\*(C'\fR to install the new binaries.
.IP "6" 4
.IX Item "6"
Run \f(CW\*(C`ovdb_init \-u\*(C'\fR as the news user.
.IP "7" 4
.IX Item "7"
Start \s-1INN\s0 with \f(CW\*(C`rc.news\*(C'\fR
.PP
It is \s-1OK\s0 to specify \f(CW\*(C`\-u\*(C'\fR even if no upgrades are needed.
.SH "HISTORY"
.IX Header "HISTORY"
Written by Heath Kehoe <hakehoe@avalon.net> for InterNetNews.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIovdb\fR\|(5), \fImakehistory\fR\|(8)


NEWS
Changes in 2.4.6

    * Fixed the segfault of the radius authenticator when none of the radius
      servers respond.  Thanks to Matija Nalis for this patch.

    * Fixed a lost initialization in buffindexed, which resolves a potential
      segfault, thanks to a patch by Kirill Berezin.

    * INN now properly supports Perl 5.10.0; Perl filters were causing innd
      to segfault on a few systems like FreeBSD.

    * Fixed a long-standing bug which affected Perl hooks for innd: the
      variable containing the body of an article was not properly created,
      which caused regular expressions matching new lines to fail.  It
      especially affected filters like Cleanfeed which sometimes failed to
      detect unwanted articles.

      To fix that issue, Julien Elie added the use of a shared string,
      available since Perl 5.7.2, with a fall back to a slower but reliable
      copy of such bodies in case the function is not available.  Using a
      Perl version superior to 5.7.2 is therefore recommended.

    * Fixed two bugs which could prevent nnrpd from being run as a daemon in
      FreeBSD.  Thanks to Johan van Selst for having identified the problem
      and to Kai Gallasch for having provided a testing FreeBSD server.  The
      listening address was not initialized to "::0" or 0.0.0.0 when the -b
      flag was not used and an incorrect size was given when IPv6 was
      enabled and the binding done using IPv4.

    * Some annoying assertion failures occurring in innfeed have been fixed
      by Russ Allbery and Julien Elie.

    * Fixed a bug in mod-active for aliased newsgroups.  Only "=" was
      written to the active file.  Thanks to D. Stussy for this patch.

    * Fixed a bug which caused innd not to honour the Ad flag in newsfeeds.

    * Fixed a bug in the IP address displayed for "localhost" in innd's
      status file.  It was not correctly initialized.

    * Fixed a permission issue:  XHDR and XPAT were not checking the rights
      the user had to read articles when accessing them by their message-ID.

    * Fixed a bug in the replies of XHDR, XOVER and XPAT when the newsgroup
      is empty.  Two initial replies were sent instead of one:  the right
      420 code followed by a wrong 224 code.

    * When no newsgroup is selected, LISTGROUP now returns the right 412
      code (instead of 481).

    * inncheck now uses a range of permissions to see whether the file modes
      are correctly set.  Therefore, different configurations depending on
      the security the user wants to enforce on his sytem are possible.

    * A new improved version of docheckgroups is shipped with INN.  The -u
      flag permits to automatically update the newsgroups file (with a
      proper number of tabulations and an alphabetical sort), removing
      obsolete descriptions and adding new ones.  A second argument on
      command-line permits to specify which newsgroups should not be
      checked, so as not to treat them.

    * An *email=* keyword has been added by James Ralston to news.daily in
      order to supply another mail address than the one set at configure
      time for Usenet daily reports.

    * An updated moderators file with information about the aioe.*, perl.*
      and si.* hierarchies is provided; control.ctl is also up to date.

    * INN supports Berkeley DB 4.7, which is the recommended version to use
      owing to various bugs affecting previous versions of Berkeley DB.

    * Other minor bugs have also been fixed.

Changes in 2.4.5

    * Fixed the "alarm signal" around "SSL_read" in nnrpd:  it allows a
      proper disconnection of news clients which were previously hanging
      when posting an article through a SSL connection.  Moreover, the
      *clienttimeout* parameter now works on SSL connections.  Thanks to
      Matija Nalis for the patch.

    * SO_KEEPALIVE is now implemented for SSL TCP connections on systems
      which support it, allowing system detection and closing the dead TCP
      SSL connections automatically after system-specified time.  Thanks to
      Matija Nalis for the patch.

    * Fixed a segmentation fault when an article of a size greater than
      remaining stack is retrieved via SSL.  Thanks to Chris Caputo for this
      patch.

    * Fixed a few segfaults and bugs which affected both Python innd and
      nnrpd hooks.  They no longer check the existence of methods not used
      by the hooked script.  An issue with Python exception handling was
      also fixed, as well as a segfault fixed by Russ Allbery which happened
      whenever one closes and then reopens Python in the same process. 
      Julien Elie also fixed a bug when reloading Python filters (they were
      not always correctly reloaded) and a segfault when generating access
      groups with embedded Python filters for nnrpd.  Many thanks to David
      Hlacik for its bug reports.

    * The nnrpd.py stub file in order to test Python nnrpd hooks, as
      mentioned in their documentation, is now installed; only INN.py was
      previously installed in *pathfilter*.  Also fixed a bug in INN.py and
      add missing methods to it.

    * Fixed a long-standing bug in innreport which prevented it from
      correctly reporting nnrpd and innfeed log messages.

    * Fixed a hang in Perl hooks on (at least) HP/PA since Perl 5.10.

    * Fixed a compilation problem on some platforms because of AF_INET6
      which was not inside a HAVE_INET6 block in innfeed.

    * Fixed a bug in innfeed which contained thrice the same IPs for each
      peer; it unnecessarily slowed the peer IP rotation for innfeed. 
      Thanks, D. Stussy, for having seen that.  Miquel van Smoorenburg
      provided the patch.

    * A new *heavily* improved version of pullnews is shipped with this INN
      release.  This new version is provided by Geraint Edwards.  He added
      no more than 16 flags, fixed some bugs and integrated the backupfeed

frontends/ovdb_init.c
/*
 * ovdb_init
 *  Performs recovery on OV database, if needed
 *  Performs upgrade of OV database, if needed and if '-u' used
 *  Starts ovdb_monitor, if needed
 */

#include "config.h"
#include "clibrary.h"
#include "inn/libinn.h"
#include <errno.h>
#include <syslog.h>

#include "inn/innconf.h"
#include "inn/messages.h"
#include "inn/ov.h"
#include "../storage/ovdb/ovdb.h"
#include "../storage/ovdb/ovdb-private.h"

#ifndef USE_BERKELEY_DB

int main(int argc UNUSED, char **argv UNUSED)
{
    die("Berkeley DB support not compiled");
}

#else /* USE_BERKELEY_DB */

static int open_db(DB **db, const char *name, int type)
{
    int ret;
    ret = db_create(db, OVDBenv, 0);
    if (ret != 0) {
	warn("db_create failed: %s\n", db_strerror(ret));
	return ret;
    }
    ret = (*db)->open(*db, NULL, name, NULL, type, DB_CREATE, 0666);
    if (ret != 0) {
	(*db)->close(*db, 0);
        warn("%s->open failed: %s", name, db_strerror(ret));
	return ret;
    }
    return 0;
}

/* Upgrade Berkeley DB version */
static int upgrade_database(const char *name UNUSED)
{
    int ret;
    DB *db;

    ret = db_create(&db, OVDBenv, 0);
    if (ret != 0)
	return ret;

    notice("upgrading %s...", name);
    ret = db->upgrade(db, name, 0);
    if (ret != 0)
        warn("db->upgrade(%s) failed: %s", name, db_strerror(ret));

    db->close(db, 0);
    return ret;
}


struct groupstats {
    ARTNUM low;
    ARTNUM high;
    int count;
    int flag;
    time_t expired;
};

static int v1_which_db(char *group)
{
    HASH grouphash;
    unsigned int i;

    grouphash = Hash(group, strlen(group));
    memcpy(&i, &grouphash, sizeof(i));
    return i % ovdb_conf.numdbfiles;
}

/* Upgrade ovdb data format version 1 to 2 */
/* groupstats and groupsbyname are replaced by groupinfo */
static int upgrade_v1_to_v2(void)
{
    DB *groupstats, *groupsbyname, *groupinfo, *vdb;
    DBT key, val, ikey, ival;
    DBC *cursor;
    group_id_t gid, higid = 0, higidbang = 0;
    struct groupinfo gi;
    struct groupstats gs;
    char group[MAXHEADERSIZE];
    u_int32_t v2 = 2;
    int ret;

    notice("upgrading data to version 2");
    ret = open_db(&groupstats, "groupstats", DB_BTREE);
    if (ret != 0)
	return ret;
    ret = open_db(&groupsbyname, "groupsbyname", DB_HASH);
    if (ret != 0)
	return ret;
    ret = open_db(&groupinfo, "groupinfo", DB_BTREE);
    if (ret != 0)
	return ret;

    memset(&key, 0, sizeof key);
    memset(&val, 0, sizeof val);
    memset(&ikey, 0, sizeof ikey);
    memset(&ival, 0, sizeof ival);

    ret = groupsbyname->cursor(groupsbyname, NULL, &cursor, 0);
    if (ret != 0)
	return ret;

    while((ret = cursor->c_get(cursor, &key, &val, DB_NEXT)) == 0) {
	if(key.size == 1 && *((char *)(key.data)) == '!') {
	    if(val.size == sizeof(group_id_t))
		memcpy(&higidbang, val.data, sizeof(group_id_t));
	    continue;
	}
	if(key.size >= MAXHEADERSIZE)
	    continue;
	memcpy(group, key.data, key.size);
	group[key.size] = 0;

	if(val.size != sizeof(group_id_t))
	    continue;
	memcpy(&gid, val.data, sizeof(group_id_t));
	if(gid > higid)
	    higid = gid;
	ikey.data = &gid;
	ikey.size = sizeof(group_id_t);

        ret = groupstats->get(groupstats, NULL, &ikey, &ival, 0);
	if (ret != 0)
	    continue;
	if(ival.size != sizeof(struct groupstats))
	    continue;
	memcpy(&gs, ival.data, sizeof(struct groupstats));

	gi.low = gs.low;
	gi.high = gs.high;
	gi.count = gs.count;
	gi.flag = gs.flag;
	gi.expired = gs.expired;
	gi.current_gid = gi.new_gid = gid;
	gi.current_db = gi.new_db = v1_which_db(group);
	gi.expiregrouppid = gi.status = 0;

	val.data = &gi;
	val.size = sizeof(gi);
        ret = groupinfo->put(groupinfo, NULL, &key, &val, 0);
	if (ret != 0) {
            warn("groupinfo->put failed: %s", db_strerror(ret));
	    cursor->c_close(cursor);
	    return ret;
	}
    }
    cursor->c_close(cursor);
    if(ret != DB_NOTFOUND) {
        warn("cursor->get failed: %s", db_strerror(ret));
	return ret;
    }

    higid++;
    if(higidbang > higid)
	higid = higidbang;

    key.data = (char *) "!groupid_freelist";
    key.size = sizeof("!groupid_freelist");
    val.data = &higid;
    val.size = sizeof(group_id_t);

    ret = groupinfo->put(groupinfo, NULL, &key, &val, 0);
    if (ret != 0) {
        warn("groupinfo->put failed: %s", db_strerror(ret));
	return ret;
    }

    ret = open_db(&vdb, "version", DB_BTREE);
    if (ret != 0)
	return ret;

    key.data = (char *) "dataversion";
    key.size = sizeof("dataversion");
    val.data = &v2;
    val.size = sizeof v2;

    ret = vdb->put(vdb, NULL, &key, &val, 0);
    if (ret != 0) {
        warn("version->put failed: %s", db_strerror(ret));
	return ret;
    }

    groupstats->close(groupstats, 0);
    groupsbyname->close(groupsbyname, 0);
    groupinfo->close(groupinfo, 0);
    vdb->close(vdb, 0);
    
    ret = db_create(&groupstats, OVDBenv, 0);
    if (ret != 0)
	return ret;
    groupstats->remove(groupstats, "groupstats", NULL, 0);
    ret = db_create(&groupsbyname, OVDBenv, 0);
    if (ret != 0)
	return ret;
    groupsbyname->remove(groupsbyname, "groupsbyname", NULL, 0);

    return 0;
}

static int check_upgrade(int do_upgrade)
{
    int ret, i;
    DB *db;
    DBT key, val;
    u_int32_t dv;
    char name[50];

    if(do_upgrade && (ret = upgrade_database("version")))
	return ret;

    ret = open_db(&db, "version", DB_BTREE);
    if (ret != 0)
	return ret;

    memset(&key, 0, sizeof key);
    memset(&val, 0, sizeof val);
    key.data = (char *) "dataversion";
    key.size = sizeof("dataversion");
    ret = db->get(db, NULL, &key, &val, 0);
    if (ret != 0) {
	if(ret != DB_NOTFOUND) {
            warn("cannot retrieve version: %s", db_strerror(ret));
	    db->close(db, 0);
	    return ret;
	}
    }
    if(ret == DB_NOTFOUND || val.size != sizeof dv) {
	dv = DATA_VERSION;

	val.data = &dv;
	val.size = sizeof dv;
        ret = db->put(db, NULL, &key, &val, 0);
	if (ret != 0) {
            warn("cannot store version: %s", db_strerror(ret));
	    db->close(db, 0);
	    return ret;
	}
    } else
	memcpy(&dv, val.data, sizeof dv);

    key.data = (char *) "numdbfiles";
    key.size = sizeof("numdbfiles");
    if ((ret = db->get(db, NULL, &key, &val, 0)) == 0)
	if(val.size == sizeof(ovdb_conf.numdbfiles))
	    memcpy(&(ovdb_conf.numdbfiles), val.data, sizeof(ovdb_conf.numdbfiles));
    db->close(db, 0);

    if(do_upgrade) {
	if(dv == 1) {
            ret = upgrade_database("groupstats");
	    if (ret != 0)
		return ret;
            ret = upgrade_database("groupsbyname");
	    if (ret != 0)
		return ret;
	} else {
            ret = upgrade_database("groupinfo");
	    if (ret != 0)
		return ret;
	}
        ret = upgrade_database("groupaliases");
	if (ret != 0)
	    return ret;
	for(i = 0; i < ovdb_conf.numdbfiles; i++) {
	    snprintf(name, sizeof(name), "ov%05d", i);
            ret = upgrade_database(name);
	    if (ret != 0)
		return ret;
	}
    }

    if(dv > DATA_VERSION_COMPRESS) {
        warn("cannot open database: unknown version %d", dv);
	return EINVAL;
    }
    if(dv < DATA_VERSION) {
	if(do_upgrade)
	    return upgrade_v1_to_v2();

        warn("database needs to be upgraded");
	return EINVAL;
    }
    return 0;
}

static int
upgrade_environment(void)
{
    int ret;

    ovdb_close_berkeleydb();
    ret = ovdb_open_berkeleydb(OV_WRITE, OVDB_UPGRADE);
    if (ret != 0)
	return ret;
    ret = OVDBenv->remove(OVDBenv, ovdb_conf.home, 0);
    if (ret != 0)
	return ret;
    OVDBenv = NULL;
    ret = ovdb_open_berkeleydb(OV_WRITE, 0);
    return ret;
}

int main(int argc, char **argv)
{
    int ret, c, do_upgrade = 0, recover_only = 0, err = 0;
    bool locked;
    int flags;

    openlog("ovdb_init", L_OPENLOG_FLAGS | LOG_PID, LOG_INN_PROG);
    message_program_name = "ovdb_init";

    if (!innconf_read(NULL))
        exit(1);

    if(strcmp(innconf->ovmethod, "ovdb"))
        die("ovmethod not set to ovdb in inn.conf");

    if(!ovdb_check_user())
        die("command must be run as runasuser user");

    chdir(innconf->pathtmp);

    while((c = getopt(argc, argv, "ru")) != -1) {
	switch(c) {
	case 'r':
	    recover_only = 1;
	    break;
	case 'u':
	    do_upgrade = 1;
	    break;
	case '?':
            warn("unrecognized option -%c", optopt);
	    err++;
	    break;
	}
    }
    if(recover_only && do_upgrade) {
        warn("cannot use both -r and -u at the same time");
	err++;
    }
    if(err) {
	fprintf(stderr, "Usage: ovdb_init [-r|-u]\n");
	exit(1);
    }

    locked = ovdb_getlock(OVDB_LOCK_EXCLUSIVE);
    if(locked) {
	if(do_upgrade) {
            notice("database is quiescent, upgrading");
	    flags = OVDB_RECOVER | OVDB_UPGRADE;
	}
	else {
            notice("database is quiescent, running normal recovery");
	    flags = OVDB_RECOVER;
	}
    } else {
        warn("database is active");
	if(do_upgrade) {
            warn("upgrade will not be attempted");
	    do_upgrade = 0;
	}
	if(recover_only)
            die("recovery will not be attempted");
	ovdb_getlock(OVDB_LOCK_ADMIN);
	flags = 0;
    }

    ret = ovdb_open_berkeleydb(OV_WRITE, flags);
    if(ret == DB_RUNRECOVERY) {
	if(locked)
            die("database could not be recovered");
	else {
            warn("database needs recovery but cannot be locked");
            die("other processes accessing the database must exit to start"
                " recovery");
        }
    }
    if(ret != 0)
        die("cannot open Berkeley DB: %s", db_strerror(ret));

    if(recover_only)
	exit(0);

    if(do_upgrade) {
	ret = upgrade_environment();
	if(ret != 0)
	    die("cannot upgrade Berkeley DB environment: %s", db_strerror(ret));
    }

    if(check_upgrade(do_upgrade)) {
	ovdb_close_berkeleydb();
	exit(1);
    }

    ovdb_close_berkeleydb();
    ovdb_releaselock();

    if(ovdb_check_pidfile(OVDB_MONITOR_PIDFILE) == false) {
        notice("starting ovdb monitor");
	switch(fork()) {
	case -1:
            sysdie("cannot fork");
	case 0:
	    setsid();
	    execl(concatpath(innconf->pathbin, "ovdb_monitor"),
		"ovdb_monitor", SPACES, NULL);
            syswarn("cannot exec ovdb_monitor");
	    _exit(1);
	}
	sleep(2);	/* give the monitor a chance to start */
    } else
        warn("ovdb_monitor already running");

    if(ovdb_conf.readserver) {
	if(ovdb_check_pidfile(OVDB_SERVER_PIDFILE) == false) {
            notice("starting ovdb server");
            daemonize(innconf->pathtmp);
            execl(concatpath(innconf->pathbin, "ovdb_server"), "ovdb_server",
                SPACES, NULL);
            syswarn("cannot exec ovdb_server");
            _exit(1);
	} else
            warn("ovdb_server already running");
    }

    exit(0);
}
#endif /* USE_BERKELEY_DB */

